doc/api/zlib.md

```diff
@@ -4,60 +4,124 @@
-  .on('error', () => {
-    // handle error
+const { createGzip } = require('zlib');
+const { pipeline } = require('stream');
+const {
+  createReadStream,
+  createWriteStream
+} = require('fs');
+
+const gzip = createGzip();
+const source = createReadStream('input.txt');
+const destination = createWriteStream('input.txt.gz');
+
+pipeline(source, gzip, destination, (err) => {
+  if (err) {
+    console.error('An error occurred:', err);
+    process.exitCode = 1;
+  }
+});
+
+// Or, Promisified
+
+const { promisify } = require('util');
+const pipe = promisify(pipeline);
+
+async function do_gzip(input, output) {
+  const gzip = createGzip();
+  const source = createReadStream(input);
+  const destination = createWriteStream(output);
+  await pipe(source, gzip, destination);
+}
+
+do_gzip('input.txt', 'input.txt.gz')
+  .catch((err) => {
+    console.error('An error occurred:', err);
+    process.exitCode = 1;
   });
 ```
 
 It is also possible to compress or decompress data in a single step:
 
 ```js
+const { deflate, unzip } = require('zlib');
+
 const input = '.................................';
-zlib.deflate(input, (err, buffer) => {
-  if (!err) {
-    console.log(buffer.toString('base64'));
-  } else {
-    // handle error
+deflate(input, (err, buffer) => {
+  if (err) {
+    console.error('An error occurred:', err);
+    process.exitCode = 1;
   }
+  console.log(buffer.toString('base64'));
 });
 
 const buffer = Buffer.from('eJzT0yMAAGTvBe8=', 'base64');
-zlib.unzip(buffer, (err, buffer) => {
-  if (!err) {
-    console.log(buffer.toString());
-  } else {
-    // handle error
+unzip(buffer, (err, buffer) => {
+  if (err) {
+    console.error('An error occurred:', err);
+    process.exitCode = 1;
   }
+  console.log(buffer.toString());
 });
+
+// Or, Promisified
+
+const { promisify } = require('util');
+const do_unzip = promisify(unzip);
+
+do_unzip(buffer)
+  .then((buf) => console.log(buf.toString()))
+  .catch((err) => {
+    console.error('An error occurred:', err);
+    process.exitCode = 1;
+  });
+```
+
+## Threadpool Usage and Performance Considerations
+
+All zlib APIs, except those that are explicitly synchronous, use Node.js'
+internal threadpool. This can lead to surprising effects and performance
+limitations in some applications.
+
+Creating and using a large number of zlib objects simultaneously can cause
+significant and unrecoverable memory fragmentation errors to occur that can
+be catastrophic to the Node.js process.
+
+This problem is illustrated in the following example:
+
+```js
+const zlib = require('zlib');
+
+const payload = Buffer.from('This is some data');
+
+// Caution: do not do this
+for (let i = 0; i < 30000; ++i) {
+  zlib.deflate(payload, (err, buffer) => {});
+}
 ```
 
-## Threadpool Usage
+In this example, 30,000 deflate instances are created concurrently. Because
+of how some operating systems handle memory allocation and de-allocation
+in this case, this may lead to to significant memory fragmentation.
```

Comment:
```suggestion
this may lead to to significant memory fragmentation.
```

This would require the comma to be moved to the previous line after `deallocation`. I'll add that to the suggestion above, but then they need to be accepted or rejected as a pair....